### PR TITLE
edit and beef up trainer guide

### PR DESCRIPTION
**Pull request by sad-**
edit and rewrite trainer guide
added summary and next steps sections

### CODE REVIEW

File: python/guide/packages/gluon/trainer.md [hunk 1]
```diff
@@ -1,72 +1,95 @@
 # Trainer
 
-A trainer updates neural network parameters by an optimization
-method.
+Training a neural network model consists of iteratively performing three simple steps. 
 
-## Basic Usages
+The first step is the forward step which computes the loss.  In MXNet gluon, this first step is achieved by doing a forward pass by calling `net.forward(X)` or simply `net(X)` and then calling the loss function with the result of the forward pass and the labels. For example `l = loss_fn(net(X), y)`.
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: MXNet gluon -> MXNet Gluon

File: python/guide/packages/gluon/trainer.md [hunk 2]
```diff
@@ -1,72 +1,95 @@
...
 
-We first create a simple Perceptron and run forward and
-backward passes to obtain the gradients.
+The second step is the backward step which computes the gradient of the loss with respect to the parameters. In gluon, this step is  achieved by doing the first step in an `autograd.record()` scope to record the computations needed to calculate the loss, and then calling `l.backward()` to compute the gradient of the loss with respect to the parameters.
+
+The final step is to update the neural network model parameters using an optimization algorithm. In gluon this step is performed by the [`gluon.Trainer` API](http://beta.mxnet.io/api/gluon/mxnet.gluon.Trainer.html) and is the subject of this guide. The gluon trainer is instantiated with a set of learnable parameters and the optimization algorithm to learn those parameters and every training iteration, `trainer.step` updates the model parameters.
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: remove word API

Comment (thomelane) [reply python/guide/packages/gluon/trainer.md]: gluon trainer -> `gluon.Trainer`

Comment (thomelane) [reply python/guide/packages/gluon/trainer.md]: Would reword last sentence: When creating a `gluon.Trainer` you must provide a collection of parameters that need to be learnt. You also provide an `Optimizer` that will be used to update the parameters every training iteration when `trainer.step` is called.

File: python/guide/packages/gluon/trainer.md [hunk 3]
```diff
@@ -1,72 +1,95 @@
...
-A trainer updates neural network parameters by an optimization
-method.
+Training a neural network model consists of iteratively performing three simple steps. 
 
-## Basic Usages
+The first step is the forward step which computes the loss.  In MXNet gluon, this first step is achieved by doing a forward pass by calling `net.forward(X)` or simply `net(X)` and then calling the loss function with the result of the forward pass and the labels. For example `l = loss_fn(net(X), y)`.
 
-We first create a simple Perceptron and run forward and
-backward passes to obtain the gradients.
+The second step is the backward step which computes the gradient of the loss with respect to the parameters. In gluon, this step is  achieved by doing the first step in an `autograd.record()` scope to record the computations needed to calculate the loss, and then calling `l.backward()` to compute the gradient of the loss with respect to the parameters.
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: Add link to autograd tutorial somewhere here.

File: python/guide/packages/gluon/trainer.md [hunk 4]
```diff
@@ -1,72 +1,95 @@
...
+
+The final step is to update the neural network model parameters using an optimization algorithm. In gluon this step is performed by the [`gluon.Trainer` API](http://beta.mxnet.io/api/gluon/mxnet.gluon.Trainer.html) and is the subject of this guide. The gluon trainer is instantiated with a set of learnable parameters and the optimization algorithm to learn those parameters and every training iteration, `trainer.step` updates the model parameters.
+
+## Basic Usage
+
+### Network and Trainer
+
+To illustrate how to use the gluon Trainer we will create a simple perceptron model and create a Trainer using the perceptron model parameters and a simple optimizer - 'sgd' with learning rate as 1.
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: 'sgd' -> `sgd`

File: python/guide/packages/gluon/trainer.md [hunk 5]
```diff
@@ -1,72 +1,95 @@
...
-loss = gluon.loss.L2Loss()
 net = gluon.nn.Dense(1)
 net.initialize()
+
+trainer = gluon.Trainer(net.collect_params(),
+                        optimizer='sgd', optimizer_params={'learning_rate':1})
+
+```
+
+The trainer constructor also accepts the following keyword arguments for :
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: ln28 - ln32: pull out into an advanced section at the end of the tutorial.

File: python/guide/packages/gluon/trainer.md [hunk 6]
```diff
@@ -1,72 +1,95 @@
...
+
+- `kvstore` – how key value store  should be created for multi-gpu and distributed training. Check out  [`mxnet.kvstore.KVStore`](http://beta.mxnet.io/api/gluon-related/mxnet.kvstore.KVStore.html#mxnet.kvstore.KVStore) for more information. String options are any of the following ['local', 'device', 'dist_device_sync', 'dist_device_async'].
+- `compression_params` – Specifies type of gradient compression and additional arguments depending on the type of compression being used. See [mxnet.KVStore.set_gradient_compression_method](http://beta.mxnet.io/api/gluon-related/_autogen/mxnet.kvstore.KVStore.set_gradient_compression.html#mxnet.kvstore.KVStore.set_gradient_compression) for more details on gradient compression.
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: mxnet.KVStore.set_gradient_compression_method -> `mxnet.KVStore.set_gradient_compression_method`

File: python/guide/packages/gluon/trainer.md [hunk 7]
```diff
@@ -1,72 +1,95 @@
...
+- `update_on_kvstore` – Whether to perform parameter updates on kvstore. If None, then trainer will choose the more suitable option depending on the type of kvstore.
+
+### Forward and Backward Pass
+
+Before we can use the trainer we have created to update model parameters, we must first run the forward and backward passes. Here we implement a function to compute the first two steps (forward step and backward step) of training the perceptron on a random dataset.
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: remove: we have created

File: python/guide/packages/gluon/trainer.md [hunk 8]
```diff
@@ -1,72 +1,95 @@
...
 def forward_backward():
     with autograd.record():
         l = loss(net(X), y)
     l.backward()
 forward_backward()
 ```
 
-Next create a `Trainer` instance with specifying the network parameters and
-optimization method, which is plain SGD with learning rate $\eta=1$.
+**Warning**: It is extremely important that the gradients of the loss function with respect to your model parameters are computed before running the trainer step. A common way to introduce bugs to your model training code is to omit the `loss.backward()`before the update step.
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: like it!

File: python/guide/packages/gluon/trainer.md [hunk 9]
```diff
@@ -1,72 +1,95 @@
...
 
 ```{.python .input}
-cur_weight = net.weight.data().copy()
+curr_weight = net.weight.data().copy()
+print(curr_weight)
 ```
 
-Now call the `step` method to perform one update. It accepts the `batch_size` as
-an argument to normalize the gradients. We can see the weight is changed.
+Now we will call the `step` method to perform one update. It accepts the `batch_size` as an argument to normalize the gradients. We can see the network parameters have now changed.
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: "normalize the gradients". Can you expand on this a little more?

File: python/guide/packages/gluon/trainer.md [hunk 10]
```diff
@@ -1,72 +1,95 @@
...
 X = nd.random.uniform(shape=(batch_size, 4))
 y = nd.random.uniform(shape=(batch_size,))
 
+loss = gluon.loss.L2Loss()
+
 def forward_backward():
     with autograd.record():
         l = loss(net(X), y)
     l.backward()
 forward_backward()
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: Add line space between definition and call

File: python/guide/packages/gluon/trainer.md [hunk 11]
```diff
@@ -75,16 +98,12 @@ trainer.step(batch_size)
 net.weight.data()
 ```
 
-For all implemented methods, please refer to the
-[API reference](http://beta.mxnet.io/api/gluon-related/mxnet.optimizer.html) for
-the `optimizer` module. Besides, the
-[Dive into Deep Learning](http://en.diveintodeeplearning.org/chapter_optimization/index.html)
-book explains each optimization methods from scratch.
+For reference and implementation details about each optimizer, please refer to the [guide](http://beta.mxnet.io/guide/packages/optimizer/optimizer.html) for the `optimizer` module. The [Dive into Deep Learning](http://en.diveintodeeplearning.org/chapter_optimization/index.html) book also has a chapter dedicated to optimization methods and explains various key optimizers in great detail.
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: Would remove reference to Dive into Deep Learning book here. And put at the end. Would link to your tutorial instead.

File: python/guide/packages/gluon/trainer.md [hunk 12]
```diff
@@ -97,6 +116,29 @@ trainer.set_learning_rate(0.1)
 trainer.learning_rate
 ```
 
-In addition, multiple pre-defined learning rate scheduling methods are
-implemented in the `mxnet.lr_scheduler` module. Please refer to [its
-tutorial](../lr_scheduler.md).
+
+
+In addition, there are multiple pre-defined learning rate scheduling methods that are already implemented in the [`mxnet.lr_scheduler`](http://beta.mxnet.io/api/gluon-related/mxnet.lr_scheduler.html) module. The learning rate schedulers can be incorporated into your trainer by passing them in as an `optimizer_param` entry. Please refer to [the lr scheduler guide](../lr_scheduler.md) to learn more.
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: [the lr scheduler guide] -> the [LR scheduler guide]
lr -> LR

File: python/guide/packages/gluon/trainer.md [hunk 13]
```diff
@@ -97,6 +116,29 @@ trainer.set_learning_rate(0.1)
...
+
+
+
+## Summary
+
+* The MXNet Gluon Trainer API is used to update the parameters of a network with a particular optimization algorithm.
+* After the forward and backward pass, the model update step is done gluon using `trainer.step()`.
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: "is done gluon using"?

File: python/guide/packages/gluon/trainer.md [hunk 14]
```diff
@@ -97,6 +116,29 @@ trainer.set_learning_rate(0.1)
...
+* A gluon Trainer can be instantiated by passing in the name of the optimizer to use and the `optimizer_params` for that optimizer or alternatively by passing in an instance of `mxnet.optimizer.Optimizer`.
+* You can change the learning rate for a gluon Trainer by setting the member variable but gluon also provides a module for learning rate scheduling.
+
+
+
+## Next Steps
+
+While optimization and optimizers play a significant role in deep learning model training, there are still other important components to model training. Here are a few suggestions about where to look next.
+
+* The [optimizer API](http://beta.mxnet.io/api/gluon-related/mxnet.optimizer.html) and [guide](http://beta.mxnet.io/guide/packages/optimizer/optimizer.html) have information about all the different optimizers implemented in MXNet and their update steps.
```
Comment (thomelane) [inline python/guide/packages/gluon/trainer.md]: [optimizer API] -> Optimizer API

File: python/guide/packages/gluon/trainer.md (hunk 3 above)
Comment (sad-) [reply python/guide/packages/gluon/trainer.md]: done in next commit

File: python/guide/packages/gluon/trainer.md (hunk 2 above)
Comment (sad-) [reply python/guide/packages/gluon/trainer.md]: addressed in next commit. resolving comment.

Comment (sad-) [reply python/guide/packages/gluon/trainer.md]: thanks. changed and addressed.

File: python/guide/packages/gluon/trainer.md (hunk 3 above)
Comment (sad-) [reply python/guide/packages/gluon/trainer.md]: will do.

File: python/guide/packages/gluon/trainer.md (hunk 4 above)
Comment (sad-) [reply python/guide/packages/gluon/trainer.md]: done.

File: python/guide/packages/gluon/trainer.md (hunk 5 above)
Comment (sad-) [reply python/guide/packages/gluon/trainer.md]: done.

File: python/guide/packages/gluon/trainer.md (hunk 10 above)
Comment (sad-) [reply python/guide/packages/gluon/trainer.md]: done.

File: python/guide/packages/gluon/trainer.md (hunk 12 above)
Comment (sad-) [reply python/guide/packages/gluon/trainer.md]: done

File: python/guide/packages/gluon/trainer.md (hunk 13 above)
Comment (sad-) [reply python/guide/packages/gluon/trainer.md]: typo rephrased.

File: python/guide/packages/gluon/trainer.md (hunk 14 above)
Comment (sad-) [reply python/guide/packages/gluon/trainer.md]: changed.

File: python/guide/packages/gluon/trainer.md (hunk 11 above)
Comment (sad-) [reply python/guide/packages/gluon/trainer.md]: done in next commit.

**Comment by thomelane** [comment]
thanks for the updates @sad-! think GitHub collapsed the feedback, so I just changed a few extra things.
